storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,414 @@
+type PruningStorer struct {
+	lock                  sync.RWMutex
+	fullArchive           bool
+	persisters            []*persisterData
+	closedPersistersPaths []string
+	cacher                storage.Cacher
+	bloomFilter           storage.BloomFilter
+	dbPath                string
+	persisterFactory      DbFactoryHandler
+	numOfEpochsToKeep     uint32
+	numOfActivePersisters uint32
+	identifier            string
+}
+
+// NewPruningStorer will return a new instance of PruningStorer without sharded directories' naming scheme
+func NewPruningStorer(args *PruningStorerArgs) (*PruningStorer, error) {
+	return initPruningStorer(args, "")
+}
+
+// NewShardedPruningStorer will return a new instance of PruningStorer with sharded directories' naming scheme
+func NewShardedPruningStorer(
+	args *PruningStorerArgs,
+	shardId uint32,
+) (*PruningStorer, error) {
+	shardIdStr := fmt.Sprintf("%d", shardId)
+	return initPruningStorer(args, shardIdStr)
+}
+
+// initPruningStorer will create a PruningStorer with or without sharded directories' naming scheme
+func initPruningStorer(
+	args *PruningStorerArgs,
+	shardIdStr string,
+) (*PruningStorer, error) {
+	var cache storage.Cacher
+	var db storage.Persister
+	var bf storage.BloomFilter
+	var err error
+
+	defer func() {
+		if err != nil && db != nil {
+			_ = db.Destroy()
+		}
+	}()
+
+	if args.NumOfActivePersisters < 1 {
+		return nil, storage.ErrInvalidNumberOfPersisters
+	}
+	if check.IfNil(args.Notifier) {
+		return nil, storage.ErrNilEpochStartNotifier
+	}
+	if check.IfNil(args.PersisterFactory) {
+		return nil, storage.ErrNilPersisterFactory
+	}
+
+	cache, err = storageUnit.NewCache(args.CacheConf.Type, args.CacheConf.Size, args.CacheConf.Shards)
+	if err != nil {
+		return nil, err
+	}
+
+	filePath := args.DbPath
+	if len(shardIdStr) > 0 {
+		filePath = filePath + shardIdStr
+	}
+	db, err = args.PersisterFactory.Create(filePath)
+
+	var persisters []*persisterData
+	persisters = append(persisters, &persisterData{
+		persister: db,
+		path:      filePath,
+	})
+
+	if reflect.DeepEqual(args.BloomFilterConf, storageUnit.BloomConfig{}) {
+		pdb := &PruningStorer{
+			identifier:            args.Identifier,
+			fullArchive:           args.FullArchive,
+			persisters:            persisters,
+			persisterFactory:      args.PersisterFactory,
+			closedPersistersPaths: make([]string, 0),
+			cacher:                cache,
+			bloomFilter:           nil,
+			dbPath:                filePath,
+			numOfEpochsToKeep:     args.NumOfEpochsToKeep,
+			numOfActivePersisters: args.NumOfActivePersisters,
+		}
+		err = pdb.persisters[0].persister.Init()
+		if err != nil {
+			return nil, err
+		}
+
+		pdb.registerHandler(args.Notifier)
+
+		return pdb, nil
+	}
+
+	bf, err = storageUnit.NewBloomFilter(args.BloomFilterConf)
+	if err != nil {
+		return nil, err
+	}
+
+	pdb := &PruningStorer{
+		identifier:            args.Identifier,
```

Comment:
duplicated code - make the PruningStorer without bloomFilter first. Test if bloom filter is needed and create then, add it to pruningStorer.

Reply:
done